src/test/java/com/google/jenkins/plugins/k8sengine/client/ContainerClientTest.java

```diff
@@ -23,189 +23,160 @@
+    ContainerClient containerClient =
+        setUpListClient(ImmutableList.of(TEST_CLUSTER, OTHER_CLUSTER), null);
+    List<Cluster> expected = initClusterList(ImmutableList.of(OTHER_CLUSTER, TEST_CLUSTER));
+    List<Cluster> response = containerClient.listClusters(TEST_PROJECT_ID, TEST_ZONE);
     assertNotNull(response);
     assertEquals(expected, response);
   }
 
   @Test
   public void testListClustersWithValidInputsWhenClustersIsNull() throws IOException {
+    ContainerClient containerClient = setUpListClient(null, null);
     List<Cluster> expected = ImmutableList.of();
-    List<Cluster> response = containerClient.listClusters(EMPTY_PROJECT_ID, TEST_ZONE);
+    List<Cluster> response = containerClient.listClusters(TEST_PROJECT_ID, TEST_ZONE);
     assertNotNull(response);
     assertEquals(expected, response);
   }
 
   @Test
   public void testListClustersEmptyWithValidProjectWithNoClusters() throws IOException {
+    ContainerClient containerClient = setUpListClient(ImmutableList.of(), null);
     List<Cluster> expected = ImmutableList.of();
-    List<Cluster> response = containerClient.listClusters(TEST_PROJECT_ID, OTHER_ZONE);
+    List<Cluster> response = containerClient.listClusters(TEST_PROJECT_ID, TEST_ZONE);
     assertNotNull(response);
     assertEquals(expected, response);
   }
 
   @Test(expected = IOException.class)
   public void testListClustersThrowsErrorWithInvalidProject() throws IOException {
-    containerClient.listClusters(ERROR_PROJECT_ID, TEST_ZONE);
+    ContainerClient containerClient = setUpListClient(null, new IOException());
+    containerClient.listClusters(TEST_PROJECT_ID, TEST_ZONE);
   }
 
-  private static List<Cluster> initExpectedClusterList(List<String> expectedClusterNames) {
+  private static List<Cluster> initClusterList(List<String> clusterNames) {
     List<Cluster> clusters = new ArrayList<>();
-    expectedClusterNames.forEach(e -> clusters.add(new Cluster().setName(e).setZone(TEST_ZONE)));
+    clusterNames.forEach(e -> clusters.add(new Cluster().setName(e).setZone(TEST_ZONE)));
     return clusters;
   }
 
-  private static Answer<Container.Projects.Zones.Clusters.Get> mockClustersGetAnswer(
-      ImmutableMap<String, List<Cluster>> projectToClusters) {
-    return invocation -> {
-      Object[] args = invocation.getArguments();
-      String projectId = (String) args[0];
-      String zone = (String) args[1];
-      String clusterName = (String) args[2];
-      if (!projectToClusters.containsKey(projectId)) {
-        throw new IOException(
-            String.format(
-                "Failed to find cluster, projectId: %s, zone: %s, cluster: %s",
-                projectId, zone, clusterName));
-      }
-
-      Optional<Cluster> cluster =
-          projectToClusters
-              .get(projectId)
-              .stream()
-              .filter(c -> c.getName().equals(clusterName) && c.getZone().equals(zone))
-              .findFirst();
-      if (!cluster.isPresent()) {
```

Comment:
I've moved away from using this kind of Answer here, because I think this depends on knowing the state of the world which is really not what the focus should be. In fact I'm changing how I think about the initial values that I set up for all of the tests away from being the "state" of anything external.

Each method that is being tested has dependencies, for this example that is the official Google client library's `Container` object that is an internal field of the `ContainerClient` that is being tested here. When you call the client library's `Container::get` method you get back an HTTP request object which you can call `execute` on and that returns either a Cluster object or throws an IOException. My main idea is that what happens in the execute call doesn't matter here, we're just concerned with the outcome of that call, and how our `ContainerClient::get` method behaves as a result.

In theory all of the values that our tests depend on could be passed in as arguments, and then our tests would verify the expected output. These mocks help us get part of the way there, in that the return values (or thrown exceptions) of the internal dependencies are explicitly defined at the beginning of each test.